crates/biome_deserialize/src/lib.rs

```diff
@@ -1,31 +1,346 @@
+///     const EXPECTED_TYPE: ExpectedType = ExpectedType::MAP;
+///
+///     fn visit_map(
+///         self,
+///         members: impl Iterator<Item = (impl DeserializableValue, impl DeserializableValue)>,
+///         range: TextRange,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self::Output> {
+///         const ALLOWED_KEYS: &[&str] = &["name"];
+///         let mut name = None;
+///         for (key, value) in members {
+///             let key_range = key.range();
+///             let Some(key) = TokenText::deserialize(key, diagnostics) else {
+///                 continue;
+///             };
+///             match key.text() {
+///                 "name" => {
+///                     name = Deserializable::deserialize(value, diagnostics);
+///                 },
+///                 _ => diagnostics.push(DeserializationDiagnostic::new_unknown_key(
+///                     key.text(),
+///                     key_range,
+///                     ALLOWED_KEYS,
+///                 )),
+///             }
+///         }
+///         Some(Person { name: name? })
+///     }
+/// }
+///
+/// enum Union {
+///     Bool(bool),
+///     Str(String),
+/// }
+///
+/// impl Deserializable for Union {
+///     fn deserialize(
+///         value: impl DeserializableValue,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self> {
+///         value.deserialize(UnionVisitor, diagnostics)
+///     }
+/// }
+///
+/// struct UnionVisitor;
+/// impl DeserializationVisitor for UnionVisitor {
+///     type Output = Union;
+///     const EXPECTED_TYPE: ExpectedType = ExpectedType::BOOL.union(ExpectedType::STR);
+///
+///     fn visit_bool(
+///         self,
+///         value: bool,
+///         range: TextRange,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self::Output> {
+///         Some(Union::Bool(value))
+///     }
+///
+///     fn visit_str(
+///         self,
+///         value: TokenText,
+///         range: TextRange,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self::Output> {
+///         Some(Union::Str(value.text().to_string()))
+///     }
+/// }
+/// ```
+pub trait DeserializationVisitor: Sized {
+    /// The type of the deserialized form of the visited value.
+    type Output;
+
+    /// The expected type of the visited value.
+    const EXPECTED_TYPE: ExpectedType;
+
+    /// The visited value is `null`.
+    ///
+    /// The default implementation appends an incorrect type diagnostic to `diagnostics`.
+    /// The expected type is retrieved from [Self::EXPECTED_TYPE].
+    fn visit_null(
+        self,
+        range: TextRange,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self::Output> {
+        debug_assert!(
+            !Self::EXPECTED_TYPE.contains(ExpectedType::NULL),
+            "This method should be implemented because the type is expected."
```

Comment:
```suggestion
            "This method should be implemented because the expected type is `null`."
```
I can see that we repeated the debug assert message. We should reword it, because "the type is expected" doesn't tell anything to me - as a consumer of this API. We can go beyond and tell the developer which type we expect in each function.